### PR TITLE
add attacher container to cephfs CSI provisioner

### LINKED ISSUE #2987: failed to mount PVC to pod using cephfs-csi
**Is this a bug report or feature request?**
* Bug Report
failed to attach pvc to pods

**Deviation from expected behavior:**
**Expected behavior:**
Mounting of PVC to pod should be successful

**How to reproduce it (minimal and precise):**

* deploy ceph-csi using operator
* create PVC using cephfs-csi
* mount PVC to pod

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Kubernetes version (use `kubectl version`): `1.13.0`
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)):

* pod descirbe output
```
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason              Age                   From                     Message
  ----     ------              ----                  ----                     -------
  Normal   Scheduled           9m58s                 default-scheduler        Successfully assigned default/csicephfs-demo-pod to worker2
  Warning  FailedMount         70s (x4 over 7m55s)   kubelet, worker2         Unable to mount volumes for pod "csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)": timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]
  Warning  FailedAttachVolume  61s (x11 over 9m43s)  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-5ebbb9b5-5f45-11e9-be2b-5254007e23ff" : attachment timeout for volume csi-cephfs-pvc-5ebbb9b5-5f45-11e9-be2b-5254007e23ff
```

logs from kubectl
```
[root@worker2 vagrant]# journalctl -xe|grep mypvc
Apr 15 06:16:37 worker2 kubelet[3835]: E0415 06:16:37.962016    3835 kubelet.go:1680] Unable to mount volumes for pod "csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)": timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]; skipping pod
Apr 15 06:16:37 worker2 kubelet[3835]: E0415 06:16:37.962103    3835 pod_workers.go:190] Error syncing pod 6d2a6439-5f45-11e9-be2b-5254007e23ff ("csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)"), skipping: timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]
Apr 15 06:18:53 worker2 kubelet[3835]: E0415 06:18:53.966157    3835 kubelet.go:1680] Unable to mount volumes for pod "csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)": timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]; skipping pod
Apr 15 06:18:53 worker2 kubelet[3835]: E0415 06:18:53.966196    3835 pod_workers.go:190] Error syncing pod 6d2a6439-5f45-11e9-be2b-5254007e23ff ("csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)"), skipping: timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]
Apr 15 06:21:08 worker2 kubelet[3835]: E0415 06:21:08.981803    3835 kubelet.go:1680] Unable to mount volumes for pod "csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)": timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]; skipping pod
Apr 15 06:21:08 worker2 kubelet[3835]: E0415 06:21:08.981844    3835 pod_workers.go:190] Error syncing pod 6d2a6439-5f45-11e9-be2b-5254007e23ff ("csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)"), skipping: timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]
Apr 15 06:23:24 worker2 kubelet[3835]: E0415 06:23:24.966423    3835 kubelet.go:1680] Unable to mount volumes for pod "csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)": timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]; skipping pod
Apr 15 06:23:24 worker2 kubelet[3835]: E0415 06:23:24.966443    3835 pod_workers.go:190] Error syncing pod 6d2a6439-5f45-11e9-be2b-5254007e23ff ("csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)"), skipping: timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]
Apr 15 06:25:41 worker2 kubelet[3835]: E0415 06:25:41.975890    3835 kubelet.go:1680] Unable to mount volumes for pod "csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)": timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]; skipping pod
Apr 15 06:25:41 worker2 kubelet[3835]: E0415 06:25:41.975933    3835 pod_workers.go:190] Error syncing pod 6d2a6439-5f45-11e9-be2b-5254007e23ff ("csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)"), skipping: timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]
Apr 15 06:27:58 worker2 kubelet[3835]: E0415 06:27:58.964396    3835 kubelet.go:1680] Unable to mount volumes for pod "csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)": timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]; skipping pod
Apr 15 06:27:58 worker2 kubelet[3835]: E0415 06:27:58.964484    3835 pod_workers.go:190] Error syncing pod 6d2a6439-5f45-11e9-be2b-5254007e23ff ("csicephfs-demo-pod_default(6d2a6439-5f45-11e9-be2b-5254007e23ff)"), skipping: timeout expired waiting for volumes to attach or mount for pod "default"/"csicephfs-demo-pod". list of unmounted volumes=[mypvc]. list of unattached volumes=[mypvc default-token-p8slv]
```

@rootfs @travisn

### PR DESCRIPTION
**Pull request by Madhu-1**
**Description of your changes:**
This PR adds the attacher container to be cephfs provisioner stateful sets and also added integration test to test cephfs CSI
**Which issue is resolved by this Pull Request:**
Resolves #2987

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../CONTRIBUTING.md#comments)

// known CI issues
[skip ci]

### CODE REVIEW

**Comment by Madhu-1** [comment]
@travisn PTAL

**Comment by Madhu-1** [comment]
@travisn tests passed for me locally

```
deployment.apps "rook-ceph-operator" deleted
2019-05-10 16:50:58.694375 I | installer: done removing the operator from namespace smoke-ns-system
2019-05-10 16:50:58.694446 I | installer: removing host data dir /home/mrajanna/workspace/src/github.com/rook/rook/tests/integration/rook-test
2019-05-10 16:50:58.705194 I | testutil: Running kubectl [create -f -]
job.batch/rook-cleanup-146d7e7c-9f64-45e5-99b3-4fbc8216499f created
2019-05-10 16:50:58.939726 I | installer: removing /home/mrajanna/workspace/src/github.com/rook/rook/tests/integration/rook-test from node test-prefix-this-is-a-very-long-hostname-master. err=<nil>
2019-05-10 16:50:58.939759 I | testutil: Running kubectl [create -f -]
job.batch/rook-cleanup-f5721346-2bd0-4154-9c53-4f9c90a4ff83 created
2019-05-10 16:50:59.167857 I | installer: removing /home/mrajanna/workspace/src/github.com/rook/rook/tests/integration/rook-test from node test-prefix-this-is-a-very-long-hostname-worker0. err=<nil>
2019-05-10 16:50:59.167892 I | testutil: Running kubectl [create -f -]
job.batch/rook-cleanup-b38e216a-e6d9-4ada-bd66-fe225cf6d67e created
2019-05-10 16:50:59.420481 I | installer: removing /home/mrajanna/workspace/src/github.com/rook/rook/tests/integration/rook-test from node test-prefix-this-is-a-very-long-hostname-worker1. err=<nil>
2019-05-10 16:50:59.434016 I | testutil: restoring hostname of node master to master
2019-05-10 16:50:59.464091 I | testutil: restoring hostname of node worker0 to worker0
2019-05-10 16:50:59.496665 I | testutil: restoring hostname of node worker1 to worker1
--- PASS: TestSmokeSuite (348.75s)
    --- PASS: TestSmokeSuite/TestBlockCSI_SmokeTest (161.22s)
PASS
ok  	github.com/rook/rook/tests/integration	348.765s
```

File: cluster/examples/kubernetes/ceph/csi/rbac/rbd/csi-provisioner-rbac.yaml [hunk 1]
```diff
@@ -37,12 +37,6 @@ rules:
   - apiGroups: [""]
     resources: ["nodes"]
     verbs: ["get", "list", "watch"]
-  - apiGroups: ["storage.k8s.io"]
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/csi/rbac/rbd/csi-provisioner-rbac.yaml]: the volumeattachments are needed for the rbd driver anymore?

Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/csi/rbac/rbd/csi-provisioner-rbac.yaml]: @travisn we have a multiple   `volumeattachment` at https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/csi/rbac/rbd/csi-provisioner-rbac.yaml#L35 and https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/csi/rbac/rbd/csi-provisioner-rbac.yaml#L41  so am  removing the duplicate

**Comment by ajarr** [comment]
> @ajarr Could you also review? thanks

looks good, Travis.

**Comment by ekuric** [comment]
Getting ```csi-cephfsplugin-provisioner.yaml``` from here , and trying to upload it fails with 
```
# oc create -f csi-cephfsplugin-provisioner.yaml 
error: error parsing csi-cephfsplugin-provisioner.yaml: error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{".Namespace":interface {}(nil)}
```

**Comment by ShyamsundarR** [comment]
> Getting `csi-cephfsplugin-provisioner.yaml` from here , and trying to upload it fails with
> 
> ```
> # oc create -f csi-cephfsplugin-provisioner.yaml 
> error: error parsing csi-cephfsplugin-provisioner.yaml: error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{".Namespace":interface {}(nil)}
> ```

These are templates, and cannot be applied directly. IOW, that .Namespace within curly braces needs to be replaced with the real namespace if you are using the YAML and oc commands directly (applies to other variables in the template as well)

**Comment by Madhu-1** [comment]
@travisn anything pending here?

**Comment by ekuric** [comment]
@ShyamsundarR ok, thanks. The problem I see is actually it does not create attacher pod(s) for csi cephfs @Madhu-1  are these created in your test?

**Comment by Madhu-1** [comment]
@ekuric yes,  attacher will run as  a container in provisioner pod

**Comment by travisn** [comment]
LGTM, we just have a CI issue (full disk) independent from this PR that we should fix before merging

**Comment by yydzhou** [comment]
@Madhu-1  Is there any plan to apply this change also into version 1.0? I am seeing very similar issue as of https://github.com/rook/rook/issues/2987 on branch 1.0 (version 1.04 actually.)

**Pull request merged**